src/MaxText/sharding.py

```diff
@@ -48,10 +50,42 @@ def maybe_shard_with_logical(inputs, logical_axes, mesh, shard_mode):
   """
   A wrapper of maybe_shard_with_name when logical axes are inputs
   """
+  if inputs is None:
+    return inputs
   named_sharding = NamedSharding(mesh, nn.logical_to_mesh_axes(logical_axes))
   return maybe_shard_with_name(inputs, named_sharding, shard_mode)
 
 
+def remove_size_one_mesh_axis(spec, mesh) -> P:
+  """
+  Removes mesh axes from a PartitionSpec (P) where the axis size is 1.
+
+  This is a common optimization to simplify sharding by excluding redundant axes.
+  Copied from jax._src.core:
+  https://github.com/jax-ml/jax/blob/db65f4548d10493613ee4155fd463a17ab5680ddc/jax/_src/core.py#L2139-L2148
```

Comment:
This link is not found? https://screenshot.googleplex.com/48Vou6KgByUuGkQ You could generate the permanent link instead.